File: packages/sdk-php/tests/NotifiqueTest.php
<?php

namespace Notifique\Tests;

use PHPUnit\Framework\TestCase;
use GuzzleHttp\Handler\MockHandler;
use GuzzleHttp\HandlerStack;
use GuzzleHttp\Psr7\Response;
use GuzzleHttp\Middleware;
use Notifique\Notifique;
use Notifique\Exception\NotifiqueApiException;

class NotifiqueTest extends TestCase
{
    private array $container = [];

    private function getMockClient(array $responses): Notifique
    {
        $mock = new MockHandler($responses);
        $history = Middleware::history($this->container);
        $handlerStack = HandlerStack::create($mock);
        $handlerStack->push($history);
        return new Notifique('test-key', 'https://api.notifique.dev/v1', ['handler' => $handlerStack]);
    }

    public function testWhatsAppSendText(): void
    {
        $client = $this->getMockClient([
            new Response(202, [], json_encode(['success' => true, 'data' => ['messageIds' => ['msg-123'], 'status' => 'QUEUED']])),
        ]);
        $resp = $client->whatsapp->sendText('instance-1', ['5511999999999'], 'Hello');
        $this->assertIsArray($resp);
        $this->assertSame(['msg-123'], $resp['data']['messageIds']);
        $this->assertSame('QUEUED', $resp['data']['status']);
    }

    public function testWhatsAppSendErrorThrows(): void
    {
        $client = $this->getMockClient([
            new Response(400, [], json_encode(['error' => 'Invalid instance'])),
        ]);
        $this->expectException(NotifiqueApiException::class);
        $client->whatsapp->sendText('x', ['123'], 'hi');
    }

    public function testWhatsAppSendWithParams(): void
    {
        $client = $this->getMockClient([
            new Response(202, [], json_encode(['success' => true, 'data' => ['messageIds' => ['m1'], 'status' => 'QUEUED']])),
        ]);
        $resp = $client->whatsapp->send('instance-1', [
            'to' => ['5511888888888'],
            'type' => 'text',
            'payload' => ['message' => 'Hi'],
        ]);
        $this->assertSame(['m1'], $resp['data']['messageIds']);
    }

    public function testWhatsAppSendWithIdempotencyKey(): void
    {
        $client = $this->getMockClient([
            new Response(202, [], json_encode(['success' => true, 'data' => ['messageIds' => ['m1'], 'status' => 'QUEUED']])),
        ]);
        $client->whatsapp->send('instance-1', [
            'to' => ['5511888888888'],
            'type' => 'text',
            'payload' => ['message' => 'Hi'],
        ], 'wa-idem-456');
        $this->assertCount(1, $this->container);
        $this->assertSame('wa-idem-456', $this->container[0]['request']->getHeaderLine('Idempotency-Key'));
    }

    public function testWhatsAppGetMessage(): void
    {
        $client = $this->getMockClient([
            new Response(200, [], json_encode(['success' => true, 'data' => ['messageId' => 'msg-1', 'status' => 'DELIVERED']])),
        ]);
        $resp = $client->whatsapp->getMessage('msg-1');
        $this->assertSame('msg-1', $resp['data']['messageId']);
        $this->assertSame('DELIVERED', $resp['data']['status']);
    }

    public function testWhatsAppListInstances(): void
    {
        $client = $this->getMockClient([
            new Response(200, [], json_encode(['success' => true, 'data' => [], 'pagination' => ['total' => 0]])),
        ]);
        $resp = $client->whatsapp->listInstances();
        $this->assertTrue($resp['success']);
    }

    public function testSmsSend(): void
    {
        $client = $this->getMockClient([
            new Response(202, [], json_encode([
                'success' => true,
                'data' => ['status' => 'QUEUED', 'count' => 1, 'smsIds' => ['sms-1']],
            ])),
        ]);
        $resp = $client->sms->send(['to' => ['5511999999999'], 'message' => 'Test']);
        $this->assertTrue($resp['success']);
        $this->assertSame(['sms-1'], $resp['data']['smsIds']);
        $this->assertSame('QUEUED', $resp['data']['status']);
    }

    public function testSmsSendWithIdempotencyKey(): void
    {
        $client = $this->getMockClient([
            new Response(202, [], json_encode([
                'success' => true,
                'data' => ['status' => 'QUEUED', 'count' => 1, 'smsIds' => ['sms-1']],
            ])),
        ]);
        $client->sms->send(['to' => ['5511999999999'], 'message' => 'Test'], 'my-idempotency-key');
        $this->assertCount(1, $this->container);
        $req = $this->container[0]['request'];
        $this->assertSame('my-idempotency-key', $req->getHeaderLine('Idempotency-Key'));
    }

    public function testSmsGet(): void
    {
        $client = $this->getMockClient([
            new Response(200, [], json_encode([
                'success' => true,
                'data' => ['smsId' => 'sms-1', 'status' => 'delivered'],
            ])),
        ]);
        $resp = $client->sms->get('sms-1');
        $this->assertSame('sms-1', $resp['data']['smsId']);
    }

    public function testSmsCancel(): void
    {
        $client = $this->getMockClient([
            new Response(200, [], json_encode([
                'success' => true,
                'data' => ['smsId' => 'sms-1', 'status' => 'cancelled'],
            ])),
        ]);
        $resp = $client->sms->cancel('sms-1');
        $this->assertTrue($resp['success']);
        $this->assertSame('cancelled', $resp['data']['status']);
    }

    public function testEmailSend(): void
    {
        $client = $this->getMockClient([
            new Response(202, [], json_encode([
                'success' => true,
                'data' => ['emailIds' => ['email-1'], 'status' => 'QUEUED', 'count' => 1],
            ])),
        ]);
        $resp = $client->email->send([
            'from' => 'noreply@example.com',
            'to' => ['u@example.com'],
            'subject' => 'Test',
            'text' => 'Body',
        ]);
        $this->assertSame(['email-1'], $resp['data']['emailIds']);
        $this->assertSame('QUEUED', $resp['data']['status']);
    }

    public function testEmailSendConvertsFromNameToCamelCase(): void
    {
        $client = $this->getMockClient([
            new Response(202, [], json_encode([
                'success' => true,
                'data' => ['emailIds' => ['e1'], 'status' => 'QUEUED', 'count' => 1],
            ])),
        ]);
        $client->email->send([
            'from' => 'noreply@example.com',
            'from_name' => 'Suporte',
            'to' => ['u@example.com'],
            'subject' => 'Test',
            'html' => '<p>Hi</p>',
        ]);
        $this->assertCount(1, $this->container);
        $body = json_decode($this->container[0]['request']->getBody()->getContents(), true);
        $this->assertArrayHasKey('fromName', $body);
        $this->assertSame('Suporte', $body['fromName']);
        $this->assertArrayNotHasKey('from_name', $body);
    }

    public function testEmailSendWithIdempotencyKey(): void
    {
        $client = $this->getMockClient([
            new Response(202, [], json_encode([
                'success' => true,
                'data' => ['emailIds' => ['email-1'], 'status' => 'QUEUED', 'count' => 1],
            ])),
        ]);
        $client->email->send([
            'from' => 'noreply@example.com',
            'to' => ['u@example.com'],
            'subject' => 'Test',
            'text' => 'Body',
        ], 'email-idem-123');
        $this->assertCount(1, $this->container);
        $this->assertSame('email-idem-123', $this->container[0]['request']->getHeaderLine('Idempotency-Key'));
    }

    public function testEmailCancel(): void
    {
        $client = $this->getMockClient([
            new Response(200, [], json_encode([
                'success' => true,
                'data' => ['emailId' => 'email-1', 'status' => 'cancelled'],
            ])),
        ]);
        $resp = $client->email->cancel('email-1');
        $this->assertTrue($resp['success']);
    }

    public function testMessagesSend(): void
    {
        $client = $this->getMockClient([
            new Response(200, [], json_encode([
                'success' => true,
                'data' => ['messageIds' => ['m1', 'm2'], 'status' => 'queued', 'count' => 2],
            ])),
        ]);
        $resp = $client->messages->send([
            'to' => ['5511999999999'],
            'template' => 'welcome',
            'variables' => ['name' => 'User'],
            'channels' => ['whatsapp', 'sms'],
            'instanceId' => 'inst-1',
        ]);
        $this->assertSame(['m1', 'm2'], $resp['data']['messageIds']);
    }

    public function testPushMessagesSend(): void
    {
        $client = $this->getMockClient([
            new Response(202, [], json_encode([
                'success' => true,
                'data' => ['status' => 'QUEUED', 'count' => 1, 'pushIds' => ['push-1']],
            ])),
        ]);
        $resp = $client->push->messages->send([
            'to' => ['device-1'],
            'title' => 'Hi',
            'body' => 'Test push',
        ]);
        $this->assertTrue($resp['success']);
        $this->assertSame(['push-1'], $resp['data']['pushIds']);
        $this->assertSame('QUEUED', $resp['data']['status']);
    }

    public function testPushMessagesSendWithIdempotencyKey(): void
    {
        $client = $this->getMockClient([
            new Response(202, [], json_encode([
                'success' => true,
                'data' => ['status' => 'QUEUED', 'count' => 1, 'pushIds' => ['push-1']],
            ])),
        ]);
        $client->push->messages->send(
            ['to' => ['device-1'], 'title' => 'Hi', 'body' => 'Body'],
            'push-idem-789'
        );
        $this->assertCount(1, $this->container);
        $this->assertSame('push-idem-789', $this->container[0]['request']->getHeaderLine('Idempotency-Key'));
    }
}

File: packages/sdk-elixir/test/notifique_test.exs
defmodule NotifiqueTest do
  use ExUnit.Case

  setup do
    client = Notifique.new("test-key")
    {:ok, client: client}
  end

  test "client initialization", %{client: client} do
    assert client.api_key == "test-key"
    assert client.base_url == "https://api.notifique.dev/v1"
  end

  describe "WhatsApp" do
    test "send_text success", %{client: client} do
      adapter = fn %Req.Request{} = req ->
        assert req.method == :post
        assert to_string(req.url) =~ "whatsapp/messages"
        body = req.options[:json]
        assert body["instanceId"] == "instance-1"
        assert body["payload"]["message"] == "Hello"
        {req, Req.Response.new(status: 200, body: %{"messageIds" => ["msg-123"], "status" => "queued"})}
      end

      {:ok, response} = Notifique.Whatsapp.send_text(client, "instance-1", "5511999999999", "Hello", adapter: adapter)
      assert response["messageIds"] == ["msg-123"]
      assert response["status"] == "queued"
    end

    test "send_text error returns status and body", %{client: client} do
      adapter = fn req ->
        {req, Req.Response.new(status: 400, body: %{"error" => "Invalid instance"})}
      end

      assert {:error, %{status: 400, body: body}} = Notifique.Whatsapp.send_text(client, "x", "1", "hi", adapter: adapter)
      assert body["error"] == "Invalid instance"
    end

    test "send with params", %{client: client} do
      adapter = fn req -> {req, Req.Response.new(status: 200, body: %{"messageIds" => ["m1"], "status" => "queued"})} end

      params = %{
        "to" => ["5511888888888"],
        "type" => "text",
        "payload" => %{"message" => "Hi"}
      }

      {:ok, resp} = Notifique.Whatsapp.send(client, "instance-1", params, adapter: adapter)
      assert resp["messageIds"] == ["m1"]
    end

    test "get_message", %{client: client} do
      adapter = fn %Req.Request{} = req ->
        assert to_string(req.url) =~ "whatsapp/messages/msg-1"
        {req, Req.Response.new(status: 200, body: %{"messageId" => "msg-1", "status" => "delivered"})}
      end

      {:ok, status} = Notifique.Whatsapp.get_message(client, "msg-1", adapter: adapter)
      assert status["messageId"] == "msg-1"
      assert status["status"] == "delivered"
    end

    test "list_instances", %{client: client} do
      adapter = fn req -> {req, Req.Response.new(status: 200, body: %{"success" => true, "data" => [], "pagination" => %{}})} end
      {:ok, list} = Notifique.Whatsapp.list_instances(client, %{}, adapter: adapter)
      assert list["success"] == true
    end
  end

  describe "SMS" do
    test "send", %{client: client} do
      adapter = fn %Req.Request{} = req ->
        assert to_string(req.url) =~ "sms/messages"
        assert req.options[:json]["to"] == ["5511999999999"]
        {req, Req.Response.new(status: 200, body: %{"success" => true, "data" => %{"smsIds" => ["sms-1"], "status" => "queued"}})}
      end

      {:ok, resp} = Notifique.Sms.send(client, %{"to" => ["5511999999999"], "message" => "Test"}, adapter: adapter)
      assert resp["data"]["smsIds"] == ["sms-1"]
    end

    test "get", %{client: client} do
      adapter = fn req -> {req, Req.Response.new(status: 200, body: %{"success" => true, "data" => %{"smsId" => "sms-1", "status" => "delivered"}})} end
      {:ok, resp} = Notifique.Sms.get(client, "sms-1", adapter: adapter)
      assert resp["data"]["smsId"] == "sms-1"
    end

    test "cancel", %{client: client} do
      adapter = fn req -> {req, Req.Response.new(status: 200, body: %{"success" => true, "data" => %{"smsId" => "sms-1", "status" => "cancelled"}})} end
      {:ok, resp} = Notifique.Sms.cancel(client, "sms-1", adapter: adapter)
      assert resp["success"] == true
      assert resp["data"]["status"] == "cancelled"
    end
  end

  describe "Email" do
    test "send", %{client: client} do
      adapter = fn req ->
        {req, Req.Response.new(status: 200, body: %{"success" => true, "data" => %{"emailIds" => ["email-1"], "status" => "queued"}})}
      end

      params = %{"from" => "noreply@example.com", "to" => ["u@example.com"], "subject" => "Test", "text" => "Body"}
      {:ok, resp} = Notifique.Email.send(client, params, adapter: adapter)
      assert resp["data"]["emailIds"] == ["email-1"]
    end

    test "cancel", %{client: client} do
      adapter = fn req -> {req, Req.Response.new(status: 200, body: %{"success" => true, "data" => %{"emailId" => "email-1", "status" => "cancelled"}})} end
      {:ok, resp} = Notifique.Email.cancel(client, "email-1", adapter: adapter)
      assert resp["success"] == true
    end
  end

  describe "Messages" do
    test "send", %{client: client} do
      adapter = fn %Req.Request{} = req ->
        assert to_string(req.url) =~ "templates/send"
        body = req.options[:json]
        assert body["template"] == "welcome"
        assert body["channels"] == ["whatsapp", "sms"]
        {req, Req.Response.new(status: 200, body: %{"success" => true, "data" => %{"messageIds" => ["m1", "m2"], "status" => "queued"}})}
      end

      params = %{
        "to" => ["5511999999999"],
        "template" => "welcome",
        "variables" => %{"name" => "User"},
        "channels" => ["whatsapp", "sms"],
        "instance_id" => "inst-1"
      }

      {:ok, resp} = Notifique.Messages.send(client, params, adapter: adapter)
      assert resp["data"]["messageIds"] == ["m1", "m2"]
    end
  end
end
